### PR TITLE
deal with complicated <br>s in SplitPOnBr

### PR DESCRIPTION
**Pull request by goose-life**

### CODE REVIEW

File: docpipe/html.py [hunk 1]
```diff
@@ -286,6 +287,24 @@ def __call__(self, context):
             br.getparent().addnext(p)
             br.getparent().remove(br)
 
+        # TODO: deal with e.g.
+        #  <p><b>[PCh1]</b><b>CHAPTER 1<br>THE INTERPRETATION OF LAWS ACT</b></p>
+        #  better
+        for br in reversed(list(context.html.xpath('.//p/b/br'))):
```
Comment (longhotsummer) [inline docpipe/html.py]: So we're looking for `br` deep inside `p` tags and breaking those onto new lines. Is it worthwhile making this more generic, eg `i`? It'd have to re-create all the tags in the tree between the `p` and the `br`. I don't think we want to support arbitrary tags (we could be deep inside divs etc.)

I think we need a test for this.

**Comment by goose-life** [comment]
@longhotsummer re not supporting arbitrary tags: we'll always be starting from a `<p>`, which shouldn't have `<div>`s or other structural tags inside them, right?

File: docpipe/html.py [hunk 2]
```diff
@@ -273,17 +286,59 @@ class SplitPOnBr(Stage):
...
+                elements.append(elem)
+            elements.append(p)
+            # now we have e.g. [<i>, <b>, <p>]
+            # the lowest-down, <i> in the example but at minimum <p>,
+            # gets the text
+            for i, elem in enumerate(elements):
+                if i == 0:
+                    elem.text = br.tail
+
+                    # when the immediate child of br is a nested tag
```
Comment (longhotsummer) [inline docpipe/html.py]:
```suggestion
                    # when the immediate sibling of br is a tag
```

File: docpipe/html.py [hunk 3]
```diff
@@ -273,17 +286,59 @@ class SplitPOnBr(Stage):
...
+                    # -- make sure <italics> ends up in the right place
+                    sibling = br.getnext()
+                    while sibling is not None:
+                        elem.append(sibling)
+                        sibling = sibling.getnext()
+
+                    # when the br falls inside an intervening tag, e.g.
+                    # <p>Text 1 <b>bold 1<br>bold 2</b> text 2</p>
+                    # make sure ' text 2' ends up in the right place
+                    elem.tail = br.getparent().tail
```
Comment (longhotsummer) [inline docpipe/html.py:325]: Can you add a test case for the (odd, but possible): `<p>first <br/> second</p>tail`

**Review by longhotsummer — APPROVED** [review]
nice! Just the suggestion for an extra test case

**Comment by longhotsummer** [comment]
> @longhotsummer re not supporting arbitrary tags: we'll always be starting from a `<p>`, which shouldn't have `<div>`s or other structural tags inside them, right?

yes, i think that's a fair assumption.

File: docpipe/html.py (hunk 3 above)
Comment (goose-life) [reply docpipe/html.py:325]: I was worried for a second there 😅

**Pull request merged**